Alerting: Add saved searches feature for alert rules page

dhalachliyski opened the pull request:
## Problem
Around 28% of the users who visit the Alert rules page interact with Search. A big big percent of the rest were not even looking for existing rules. Many users are only responsible for a specific set of alert rules and so they need to enter the same search criteria every time (or have a few of them which they would need to go to).

## Solution
This PR introduces the concept of Saved searches. Users can chose to save a given search string and apply it easily afterwards. They are also able to mark a search string as "default", so it is pre-applied when they visit the page.

##Video

https://github.com/user-attachments/assets/369add15-5bf3-4a87-9873-2de66a6edff5




- Add ability to save, rename, delete, and apply search queries on the Alert Rules page
- Includes auto-apply default search on navigation and UserStorage persistence

Behind feature toggle `alertingSavedSearches` (disabled by default).

## Test plan
- [ ] Enable feature toggle `alertingSavedSearches` 
- [ ] Save a search query and verify it persists
- [ ] Rename and delete saved searches
- [ ] Set a default search and verify auto-apply on navigation
- [ ] Run unit tests: `yarn test SavedSearches`
- [ ] Run e2e tests: `yarn e2e:playwright alerting-suite/saved-searches`

dhalachliyski committed:
Alerting: Add saved searches feature for alert rules page

Add ability to save, rename, delete, and apply search queries on the
Alert Rules page. Includes auto-apply default search on navigation
and UserStorage persistence.

Behind feature toggle `alertingSavedSearches` (disabled by default).

dhalachliyski committed:
Alerting: Add i18n translations for saved searches

dhalachliyski committed:
Alerting: Remove unused imports in saved searches

dhalachliyski committed:
Alerting: Add CODEOWNERS for e2e-playwright/alerting-suite

dhalachliyski committed:
Alerting: Add useSavedSearches mock to RulesFilter.v2 tests

dhalachliyski committed:
Alerting: Fix failing unit tests for saved searches

- Fix Jest mock hoisting issue in useSavedSearches.test.ts by configuring
  UserStorage mock implementation after imports instead of inline
- Update SavedSearches.test.tsx to use findBy* queries for async popup content
- Fix tests to click apply button instead of text for applying searches
- Update maxLength test to verify attribute instead of trying to exceed it

dhalachliyski committed:
Alerting: Fix saved searches test mocking and assertions

- Fix UserStorage mock in useSavedSearches.test.ts by creating mock with
  default Promise-returning functions inside jest.mock() factory, then
  accessing the instance via getMockUserStorageInstance() helper
- Fix SavedSearches.test.tsx apply button tests to use correct accessible
  name "Apply this search" (from tooltip) instead of dynamic aria-label
- Fix disabled button assertion to check native disabled attribute instead
  of relying on aria-disabled which is set inconsistently by Button component
- Use findAllByRole for async popup content queries

dhalachliyski committed:
Alerting: Fix test query for disabled save button

Use findByText + closest instead of findByRole to find the disabled
"Save current search" button. The Grafana Button component renders
with conflicting accessibility attributes (disabled="" + aria-disabled="false")
which breaks role-based queries in React Testing Library.

dhalachliyski committed:
fix(alerting): preserve UserStorage mock reference before clearAllMocks

dhalachliyski committed:
fix(alerting): add missing test mocks for crypto and console

- Mock crypto.randomUUID for Node.js test environment
- Add console.error spy to tests expecting storage/parse errors
- Add console.warn spy to test expecting validation warnings

Fixes jest-fail-on-console failures and crypto.randomUUID TypeError.

dhalachliyski committed:
fix(alerting): add console.error spy to save failure test

laurenashleigh commented on public/app/features/alerting/unified/rule-list/filter/RulesFilter.v2.tsx:
```diff
@@ -97,6 +115,39 @@ export default function RulesFilter({ viewMode, onViewModeChange }: RulesFilterP
+    }
+
+    const defaultSearch = getAutoApplySearch();
+    if (defaultSearch) {
```
Best to wrap this in a try-catch to catch any errors, e.g.
```
 const defaultSearch = getAutoApplySearch();
    if (defaultSearch) {
      try {
        handleApplySearch(defaultSearch);
      } catch (error) {
        console.error('Failed to auto-apply default search:', error);
      }
    }
```

laurenashleigh reviewed (changes requested):
Overall pretty neat! I've left some comments, it would also be good to split some of these files into smaller ones e.g. `SavedSearches.tsx` `SavedSearchItem.tsx` `InlineSaveInput.tsx` `InlineRenameInput.tsx`

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/RulesFilter.v2.tsx:
```diff
@@ -97,6 +115,39 @@ export default function RulesFilter({ viewMode, onViewModeChange }: RulesFilterP
+      return;
+    }
+
+    const defaultSearch = getAutoApplySearch();
```
This looks a bit fishy. Why the double apply can happen? I wonder if we could come up with a better condition for applying the default search params

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/SavedSearches.test.tsx:
I wonder if these tests provide much value. They tests very detailed implementation details rather than the component behaviour. I think I'd try to limit the number of tests and focus only on testing behavior and functionality rather than implementation details

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/SavedSearches.tsx:
```diff
@@ -0,0 +1,956 @@
+      return () => clearTimeout(timer);
+    }
+    return undefined;
+  }, [isOpen, activeAction]);
```
Do we need this?

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/SavedSearches.tsx:
```diff
@@ -0,0 +1,956 @@
+  const handleCancelRename = useCallback(() => {
+    setActiveAction('idle');
+  }, []);
+
```
If we want to have so detailed state tracking, it could be worth considering adding a reducer to handle the state management

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/useSavedSearches.ts:
```diff
@@ -0,0 +1,424 @@
+        throw error;
+      }
+    },
+    [notifyApp]
```
Same here. We could use `useAsync` helpers

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/useSavedSearches.ts:
```diff
@@ -0,0 +1,424 @@
+        });
+      } catch {
+        // Error already handled in persistSearches
+      }
```
We should avoid swallowing errors

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/useSavedSearches.ts:
```diff
@@ -0,0 +1,424 @@
+        // Track analytics
+        trackSavedSearchRename();
+      } catch {
+        // Error already handled in persistSearches
```
We swallow an error

dhalachliyski committed:
fix(alerting): address PR review feedback for saved searches

- Register alertingSavedSearches feature toggle in backend
- Extract shared types to SavedSearches.types.ts to fix circular dependencies
- Extract sub-components: InlineSaveInput, InlineRenameInput, SavedSearchItem
- Remove unused imports (IconButton, Input) and styles from SavedSearches.tsx
- Add try/catch for auto-apply default search error handling
- Remove maxLength validation and corresponding test

dhalachliyski committed:
fix(alerting): fix validation error display in saved searches

- Fix useEffect dependency array that was immediately clearing validation errors
- Remove error from deps so errors only clear when user types, not when set
- Run i18n-extract to remove unused error-name-too-long translation key

dhalachliyski committed:
fix(alerting): address PR review feedback for saved searches

- Replace toHaveBeenCalled assertions with UI verification using AppNotificationList
- Rename useSavedSearches.test.ts to .tsx for JSX support
- Update README documentation to reflect current test patterns
- Add test cleanup between E2E tests to prevent data leakage

dhalachliyski committed:
fix(alerting): remove unused import and fix test wrapper

- Remove unused locationService import from RulesFilter.v2.tsx
- Add missing bootData spread in useSavedSearches.test.tsx mock
- Add createWrapper to renderHook call for user-specific storage key test

dhalachliyski committed:
fix(alerting): add Redux wrapper to all useSavedSearches hook tests

All renderHook calls for useSavedSearches now include the createWrapper()
which provides the Redux Provider context required by useAppNotification.

dhalachliyski committed:
fix(alerting): use regex patterns in MSW handlers for UserStorage tests

MSW handlers now use regex patterns to match any namespace and user UID,
since UserStorage reads config values from internal imports that aren't
affected by jest.mock of @grafana/runtime.

dhalachliyski committed:
fix(alerting): mock UserStorage directly instead of using MSW

Replace MSW HTTP handlers with a direct mock of the UserStorage class.
The MSW approach failed because UserStorage evaluates config.namespace
at module load time, before jest.mock takes effect, causing the regex
patterns to not match the actual request URLs.

This follows the same pattern used in useFavoriteDatasources.test.ts.

dhalachliyski committed:
refactor(alerting): use react-hook-form and Dropdown for saved searches

- Migrate InlineRenameInput and InlineSaveInput to react-hook-form
- Replace custom PopupCard with Grafana Dropdown component
- Use useReducer for centralized dropdown state management
- Add stopPropagation handlers to prevent dropdown closing during form interactions
- Update tests to use real useSavedSearches hook with mocked UserStorage
- Consolidate and simplify saved searches test suite

dhalachliyski committed:
fix: resolve CI failures in SavedSearches component

- Fix TypeScript TS2540 errors by using MutableRefObject type for refs
- Fix form submission by using onClick instead of type="submit" on IconButton
  (IconButton doesn't forward the type prop to the underlying button)
- Fix action menu tests by stopping click propagation on ActionMenu wrapper
- Fix Escape key handling by focusing the dialog element instead of the
  potentially-disabled save button

dhalachliyski committed:
fix(alerting): add navTree to runtime mock in useSavedSearches tests

Add empty navTree array to the @grafana/runtime config mock to prevent
store initialization crash when buildInitialState() calls .find() on
undefined navTree.

dhalachliyski committed:
fix(alerting): add error handling for auto-apply default search

Wrap handleApplySearch call in try-catch to prevent unhandled exceptions
when auto-applying the default saved search on navigation.

dhalachliyski committed:
fix(alerting): prevent saved searches dropdown from closing when clicking action menu

The nested Dropdown components caused the outer SavedSearches dropdown to close
when clicking on action menu items (Set as default, Rename, Delete). This happened
because @floating-ui/react's useDismiss hook detected clicks on the inner Menu
(rendered via Portal) as "outside" clicks.

Fix: Replace the outer Dropdown with PopupCard and add custom click-outside
handling that explicitly excludes portal elements ([role="menu"] and
[data-popper-placement]). This matches the pattern used before the Dropdown
refactor.

Changes:
- SavedSearches.tsx: Use PopupCard instead of Dropdown, add click-outside handler
- SavedSearchItem.tsx: Add menuPortalRoot prop for action menu positioning
- RulesFilter.v2.tsx: Fix double analytics tracking on auto-apply

dhalachliyski committed:
fix(alerting): auto-apply default saved search on page navigation

The default saved search was not being applied when navigating to the
Alert rules page. This was caused by a race condition where `isLoading`
was `false` on initial render (status was 'not-executed'), causing the
auto-apply effect to run before saved searches were loaded.

Fix: Include the uninitialized state in the loading check so the effect
waits until data is actually loaded before attempting to auto-apply.

Also adds tests for the auto-apply functionality.

dhalachliyski committed:
fix(alerting): align action menu icon and improve saved search tests

- Fix vertical alignment of three-dot menu icon in saved search items
  by adding flex centering to the wrapper div
- Add feature toggle setup/teardown in saved searches test suite
- Fix location mocking in test for URL search parameter handling

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/InlineSaveInput.tsx:
```diff
@@ -0,0 +1,134 @@
+    if (validationError) {
+      setError('name', { message: validationError });
+      return;
+    }
```
I think we could use `react-hook-form` validation instead of doing it manually https://www.react-hook-form.com/api/useform/register

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/SavedSearches.test.tsx:
```diff
@@ -0,0 +1,353 @@
+  setAsDefaultMenuItem: byText(/set as default/i),
+  removeDefaultMenuItem: byText(/remove default/i),
+  renameMenuItem: byText(/rename/i),
+  deleteMenuItem: byText(/^delete$/i),
```
Cannot we use `byRole` or `byLabelText` selectors for menu items?

konrad147 commented on public/app/features/alerting/unified/rule-list/filter/useSavedSearches.ts:
```diff
@@ -0,0 +1,380 @@
+            'Your saved searches could not be loaded. Please try refreshing the page.'
+          )
+        );
+      });
```
I think all of these should be handled in `loadSavedSearches` function. We shouldn't mix async/await and then/catch

dhalachliyski commented on public/app/features/alerting/unified/rule-list/filter/RulesFilter.v2.test.tsx:
```diff
@@ -429,4 +467,109 @@ describe('RulesFilterV2', () => {
+      mockStorageData.savedSearches = JSON.stringify([defaultSavedSearch]);
+
+      // Set session storage to simulate a page refresh (already visited)
+      sessionStorage.setItem('grafana.alerting.alertRules.visited.123', 'true');
```
No, even if you clear it, if you navigate away and come back, you'd still see the default

dhalachliyski committed:
refactor(alerting): improve saved searches validation and organization

- Rename SavedSearches.types.ts to savedSearchesSchema.ts
- Use react-hook-form's built-in validation instead of manual setError
- Change error handling to throw ValidationError instead of returning it
- Add type guard isValidationError for safe error checking
- Add alphabetical sorting for saved searches (default first)
- Replace console.warn/error with logWarning/logError for analytics
- Extract helper functions: sortSavedSearches, loadSavedSearchesFromStorage, hasUrlSearchQuery

dhalachliyski committed:
refactor(alerting): address PR review comments for saved searches (steps 9-12)

- Add comprehensive comment explaining useEffect double-render limitation
  and potential future improvements for default search auto-apply (step 9)
- Add test documenting expected behavior when navigating back to alert list
  after leaving the page - default filter is re-applied (step 10)
- Update RulesFilter.v2.test.tsx to use testWithFeatureToggles helper and
  add MSW UserStorage handlers for future use (step 11)
- Update SavedSearches.test.tsx to use render from test/test-utils and
  byRole selectors for menu items (step 12)

dhalachliyski committed:
test(alerting): update saved searches tests for refactored API

- Update mockSavedSearches order to match sorted output (default first, then alphabetically)
- Change validation error tests to use rejects pattern (saveSearch/renameSearch now throw)
- Add hasPermission mock to contextSrv for module-level permission check

dhalachliyski committed:
fix(alerting): fix CI failures for saved searches

- Update onRenameComplete type to match throw-based API (Promise<void>)
- Run i18n-extract to add missing translation keys

dhalachliyski committed:
fix(alerting): salvage valid entries when saved searches validation fails

Instead of returning an empty array when array validation fails,
iterate through each item and keep only the valid entries.
This prevents losing all saved searches if a single entry is corrupted.

dhalachliyski committed:
test(alerting): update test to expect valid entries to be preserved

Update the test assertion to match the new behavior where valid saved
search entries are preserved when some entries fail validation, rather
than discarding all entries.

dhalachliyski committed:
fix(alerting): eliminate double API request on saved search auto-apply

Move saved searches loading and auto-apply logic from RulesFilterV2 to
RuleListPage. This ensures the default search filter is applied BEFORE
FilterView mounts, preventing double API requests on initial page load.

- Load saved searches at RuleListPage level
- Gate RuleList rendering until saved searches are loaded
- Pass savedSearchesResult as prop to avoid duplicate hook calls
- Remove auto-apply tests from RulesFilter.v2.test.tsx (behavior moved)

dhalachliyski committed:
fix(alerting): mock useSavedSearches in RuleList.v2 tests

The useSavedSearches hook triggers async state updates that complete
after tests finish, causing React act() warnings. Mock the hook to
prevent async operations during tests.

dhalachliyski committed:
refactor(alerting): migrate saved searches tests to use MSW

Address code review feedback by migrating UserStorage tests from
jest.mock to MSW-based mocking:

- Add MSW helper functions (setAlertingStorageItem, getAlertingStorageItem)
  to simplify test setup for UserStorage
- Migrate useSavedSearches.test.tsx to use MSW handlers instead of
  jest.mock('@grafana/runtime/internal')
- Migrate RulesFilter.v2.test.tsx to use MSW handlers
- Update README documentation to accurately reflect how tests use MSW
- Add tests for default search auto-apply behavior in RuleListPage
- Simplify comments to be concise and accurate

dhalachliyski committed:
fix(alerting): mock UserStorage directly in useSavedSearches tests

The UserStorage class caches its storage spec at the instance level,
and the useSavedSearches hook creates the instance at module level.
This caused test isolation issues where cached state leaked between
tests, making all tests that depended on loading data fail.

Fix by mocking UserStorage class directly instead of relying on MSW
handlers. This gives each test explicit control over what getItem
and setItem return, ensuring proper isolation.

Also update persistence assertions to verify mock.setItem calls
instead of reading from MSW storage (which the mock bypasses).

dhalachliyski committed:
refactor(alerting): remove setup helper in SavedSearches tests

Replace the `setup()` helper function with direct `render()` calls
as suggested in PR review. This makes tests more explicit about
what component is being rendered and with what props.

laurenashleigh reviewed (approved):
Awesome first PR!!! 👏🏼 LGTM 🚀

konrad147 committed:
refactor(alerting): extract default search auto-apply into dedicated hook

Moves the default saved search auto-apply logic from useSavedSearches into
a new useApplyDefaultSearch hook. This improves separation of concerns by
keeping useSavedSearches focused on CRUD operations while the new hook
handles the page-level auto-apply behavior.

Key changes:
- Created useApplyDefaultSearch hook with session-based visit tracking
- Removed getAutoApplySearch method and user-specific session keys from useSavedSearches
- Exported loadDefaultSavedSearch utility for independent default search loading
- Simplified test mocks to use loadDefaultSavedSearch instead of full hook mocking
- Removed unused savedSearchesResult prop passing through component tree

konrad147 commented:
@claude can you review this PR?

konrad147 committed:
fix(alerting): improve default search auto-apply timing and test reliability

Replace react-use's auto-executing useAsync with internal useAsync hook
for better control over when default search is loaded. This prevents
race conditions and ensures the async operation only executes when needed.

Test improvements:
- Add proper session storage cleanup in beforeEach
- Use waitFor to handle async operations correctly
- Prevent visited flag from affecting subsequent tests
- Clear mock call history between tests

The internal useAsync hook doesn't auto-execute on mount, allowing us to
control exactly when the default search loads based on conditions rather
than relying on dependency array triggers.

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/alert-rules-saved-searches

konrad147 reviewed (approved):
LGTM!

torkelo commented on public/app/features/alerting/unified/rule-list/filter/SavedSearches.tsx:
```diff
@@ -0,0 +1,556 @@
+
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
```
@dhalachliyski this should use useClickAway (or better useFloating). 

Or use Dropdown component for the actual dropdown. 

the code mentions popper 

```
const isPortalClick =
            event.target.closest('[data-popper-placement]') || event.target.closest('[role="menu"]');

```

We do not use popper anymore, but useFloating
